### PR TITLE
Design to dynamically optimize the distributed training.

### PR DESCRIPTION
**Pull request by workingloong**
### What changes were proposed in this pull request?

Design to dynamically optimize the distributed training.

### Why are the changes needed?

Design to dynamically optimize the distributed training.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Not need and it is a document.

### CODE REVIEW

File: docs/design/dynamic-optimization.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+One pain point in model training is to configure training parameters and
+strategy to to fully utilize the hardware to improve throughput. The
+configuration contains:
+
+1. perfarmance-relevant parameters such as micro-batch size and the strategy of
+activation checkpoint.
+
+1. the strategy to split model such as the split of pipelining stage and
+the FDSP unit.
```
Comment (merlintang) [inline docs/design/dynamic-optimization.md:16]: FSDP? give a link here?

File: docs/design/dynamic-optimization.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+```
+
+
+### Dynamic Adjustment Extensions
+
+To dynamically adjust the training configured strategy, the user need to use
+the elastic training extension in the training program. The elastic extension
+can adjust the configured parameters or paralllelism strategy when the config
+file is updated. Now, we will implement the following extensions:
```
Comment (merlintang) [inline docs/design/dynamic-optimization.md:115]: when the config file is updated ? what is the meaning for this ? do we need provide a strategy to make sure when the update is excuted.

Comment (workingloong) [reply docs/design/dynamic-optimization.md:115]: We may set a version value in the config. The job master will increment the version if a new configuration is generated.

File: docs/design/dynamic-optimization.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+- Hardware checker collects hardware characteristics which includes the CPU
+cores, RAM, GPU and network bandwidth of the node.
+
+- Model analyzer collects the number of floating point operations, memory
+consumption, activation memory of the model.
+
+- Training monitor collects the usage of CPU cores, RAM, GPU kernels and CUDA
+memory.
+
+- Config tuner generates a config file to notify the training process to
```
Comment (samplise) [inline docs/design/dynamic-optimization.md:90]: `Configure Tuner` or `ConfigTuner`

File: docs/design/dynamic-optimization.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+### Dynamic Adjustment Extensions
+
+To dynamically adjust the training configured strategy, the user need to use
+the elastic training extension in the training program. The elastic extension
+can adjust the configured parameters or paralllelism strategy when the config
+file is updated. Now, we will implement the following extensions:
+
+- Profiler automatically analyzes the timeline to detect the bottleneck of
```
Comment (samplise) [inline docs/design/dynamic-optimization.md:116]: `Detector` seems a little more accurate than `Profiler`.

**Review by merlintang — APPROVED** [review]

File: docs/design/dynamic-optimization.md (hunk 3 above)
Comment (workingloong) [reply docs/design/dynamic-optimization.md:90]: Configure is a verb.

File: docs/design/dynamic-optimization.md (hunk 4 above)
Comment (workingloong) [reply docs/design/dynamic-optimization.md:116]: `Detector` is only a component of `Profiler`.

**Pull request merged**